### ISSUE TITLE
Error packing Numpy scalars with struct.pack (migrated from Trac #1110)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2663]
Original ticket http://projects.scipy.org/numpy/ticket/1110
Reported 2009-05-19 by trac user bryancatanzaro, assigned to unknown.

Hello -
I'm using NumPy with PyCUDA.  PyCUDA needs to convert NumPy scalars into C structs, and does so using the standard struct.pack. 
This seemed to work with earlier versions of NumPy & Python, but with NumPy version 1.3.0 on Python 2.6.2, it causes an error when packing a NumPy scalar into an unsigned struct element.

Here is a code snippet:

```
import numpy
import struct
struct.pack('I', numpy.uint32(0))
```

And here is the error:
    sys:1: DeprecationWarning: struct integer overflow masking is deprecated
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    SystemError: Objects/longobject.c:336: bad argument to internal function

The pack proceeds correctly if I don't pack to an unsigned struct element, e.g.:
    struct.pack('i', numpy.uint32(0))

Thanks,
bryan

**Comment by thouis** [comment]
Comment in Trac by atmention:inducer, 2009-05-19

CC me

**Comment by thouis** [comment]
Comment in Trac by atmention:inducer, 2009-05-19

Hm, that didn't work--what do I need to do to be on the CC line?

**Comment by thouis** [comment]
Comment in Trac by atmention:inducer, 2009-05-19

Never mind, it dit work. :) Sorry for the spam.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2009-05-19

Works for me with python 2.5.1, numpy 1.4.0.dev6858 on 64 bit linux:

```
In [2]: struct.pack("I", uint32(0))
Out[2]: '\x00\x00\x00\x00'
```

Could you supply more information? This might also be a bug in in the struct module.

**Comment by thouis** [comment]
Comment in Trac by trac user bryancatanzaro, 2009-05-19

Replying to [comment:4 charris]:

> Works for me with python 2.5.1, numpy 1.4.0.dev6858 on 64 bit linux:
> 
> Could you supply more information? This might also be a bug in in the struct module.

Yes, it seems to be ok on python 2.5, but not on python 2.6.  Unfortunately, I have to move to python 2.6 for some other reasons.

**Comment by thouis** [comment]
Comment in Trac by atmention:inducer, 2009-11-25

Can't reproduce on Py2.6.1 with numpy 1.3.0 on x86_64.

**Comment by thouis** [comment]
Comment in Trac by trac user jbenni, 2009-12-26

Release 1.4.0rc1 with Python 2.6.4, the above still fails.  (On a Mac)

Any suggestions?

**Comment by thouis** [comment]
Comment in Trac by trac user caleb, 2010-02-03

A similar problem occurs with 1.4.0 (and 1.3.0) on Python 2.6.2 (r262:71605, Apr 14 2009, 22:40:02) [MSC v.1500 32 bit (Intel)] on win32. (I'm not using PyCUDA.)

Reproduce from the command line:

```
>>> import numpy
>>> import struct
>>> a = numpy.array([1,2,3], "uint16")
>>> struct.pack("<3H", *a)
sys:1: DeprecationWarning: struct integer overflow masking is deprecated
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: ..\Objects\longobject.c:336: bad argument to internal function
```

The same will occur with unit32 and "<3I".

Note the following works:

```
>>> struct.pack("3H", *a)
'\x01\x00\x02\x00\x03\x00'
>>> struct.pack("<3H", *(a.tolist()))
'\x01\x00\x02\x00\x03\x00'
>>> a = numpy.array([1,2,3], "int16")
>>> struct.pack("<3h", *a)
'\x01\x00\x02\x00\x03\x00'
```

and so does int32.

**Comment by thouis** [comment]
Comment in Trac by trac user zhuj, 2010-03-14

I am having the same problem with packing on Mac OS X 10.5.8. The python distribution and numpy version I am using comes with Enthought Python Distribution Python 2.5.4 |EPD 5.1.0| (r254:67916, Jul 20 2009, 15:41:21) and numpy 1.3.0. Hopefully some developer has the same version installed and could track down the problem.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2010-10-24

It's a bug in the `struct` module, and not possible to work around. `get_wrapped_ulong` should call `get_pylong` to invoke the `__index__` method for non-PyLong objects, but it doesn't.

Seems to be fixed in Python 2.7.
